### ISSUE TITLE
ReferenceError: Headers is not defined

### ISSUE THREAD

**Issue by backnight** [dipscope/JsonApiEntityProvider.TS#1]
**Describe the bug**

Library not working when trying to use it.

"ReferenceError: Headers is not defined".
[The issue is located in "src/json-api-connection.ts"](https://github.com/dipscope/JsonApiEntityProvider.TS/blob/main/src/json-api-connection.ts#L36)

The file is missing an import I think

**To Reproduce**

 - Try to use the library :D

**Comment by backnight** [comment]
The *Request* class is missing also I think

**Comment by backnight** [comment]
From what I found, it's because you developed this library with a browser while I'm trying to use it server side with NodeJS.

The Headers and Request classes are available by default in browsers but for NodeJS you need the "node-fetch" package.

I tried to add it and then add the needed import but webpack started to shit all over the place when I tried to build the library.

If you want to have a look, you can find my work there: https://github.com/backnight/JsonApiEntityProvider.TS/commit/1665df3d4eeb81af251b1fa63e5a0a3bb66f5b9a

**Comment by dpimonov** [comment]
Thanks for the first issue and your investigation. 

The point is that in NodeJS v17 and v18 the fetch API is provided. In v17 it is present as experemental feature and in v18 it is enabled out of the box, so I was not sure if I have to handle node-fetch case in the initial release.

Actually I've played a bit with it before the release and there are some tricky parts when you try to combine this with a browser based fetch in one package. You can take a state which works for node as a tmp solution. Here is a file of your interest. 

https://github.com/dipscope/JsonApiEntityProvider.TS/blob/02cfc2c0eace37a9bd1278ac9b5081a64d729a27/src/json-api-connection.ts

Most probably you have to change the webpack library target. If I remember correctly it may not compile properly into UMD module because of node-fetch. 

I'm going to think how to handle the use case for stable versions of node to provide a proper fix. There is a plan to try out [Axios](https://axios-http.com/) to avoid using direct fetch at all.

If you have an experience with a combination of node-fetch and browser fetch in one package under UMD module I'll appreciate your help or examples.

**Comment by backnight** [comment]
Hello,

Noted for Node v17 and v18! I'm using the current LTS version FYI, so... v16 :) and I think it would be a good idea to have your library supporting the current LTS if you want to be integrated into production ready projects.

No XP on my side for combining browser fetch and node-fetch, sorry :/

**Comment by dpimonov** [comment]
Hello,

Thanks for the feedback. I'm going to provide a fix for this next week. Looks like axios library fits well but I have to try it. ;)

Btw which JSON:API implementation are you using as a backend service? As you may already know filtering and pagination strategies may differ and here some extension points may be required to full fill your use cases.

**Comment by backnight** [comment]
Thanks, I will monitor the new release next week then! The issue mentioned above blocks me from finishing testing your library.

Axios is nice, I used it a lot and nothing to complain so far. Have a try!

I can't tell you about the JSON:API implementation I'm currently using because I'm still testing libraries and frameworks for my new side project. I'm not integrating your library into an existing project so I don't have any requirements (except being compatible with LTS versions :P).
So far I'm trying NestJS for my backend APIs.

**Comment by backnight** [comment]
https://github.com/nestjs/nest/issues/3566

**Comment by dpimonov** [comment]
Pushed a fix and created a new version. Give it a try!

I've ended up using [cross-fetch](https://github.com/lquixada/cross-fetch) library as it provides exact same API with minimum efforts. Now it should work for LTS version of node.

Btw if you need sample code for basic use cases you can find it in spec file not to write it from scratch.

https://github.com/dipscope/JsonApiEntityProvider.TS/blob/main/spec/json-api-entity-provider.spec.ts

Also I've checked nestjs implementation. It is not so obvious how they handle filtering. Spec is agnostic about strategies used by a server so most probably it will differ from JSON.NET implementation which is used as a default one. Such cases are considered in the library and we can discuss it in a separate issue if this will be the case.

**Comment by backnight** [comment]
Hello,

Thanks for the reactivity and the fix.

For now I'm just trying to make the library request a GET listing endpoint returning JSON, i.e: `/exchanges` in order to fetch my exchanges.

The build is working and when I run the script, it outputs 300Kb of code in my terminal (wtf?) + an error message:
```
TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at d (/home/ERASED/Projects/ERASED/api-lib/node_modules/@dipscope/json-api-entity-provider/index.js:1:32372)
    at e.createDocumentObjectEntityCollection (/home/ERASED/Projects/ERASED/api-lib/node_modules/@dipscope/json-api-entity-provider/index.js:1:36548)
    at e.createDocumentObjectPaginatedEntityCollection (/home/ERASED/Projects/ERASED/api-lib/node_modules/@dipscope/json-api-entity-provider/index.js:1:36136)
    at e.<anonymous> (/home/ERASED/Projects/ERASED/api-lib/node_modules/@dipscope/json-api-entity-provider/index.js:1:50814)
    at /home/ERASED/Projects/ERASED/api-lib/node_modules/@dipscope/json-api-entity-provider/index.js:1:45408
    at Object.next (/home/ERASED/Projects/ERASED/api-lib/node_modules/@dipscope/json-api-entity-provider/index.js:1:45513)
    at m (/home/ERASED/Projects/ERASED/api-lib/node_modules/@dipscope/json-api-entity-provider/index.js:1:44227)

```

**Comment by backnight** [comment]
My app code is really simple, to give you an overview:
```
import { JsonApiEntityProvider, JsonApiEntityProviderOptions } from '@dipscope/json-api-entity-provider';
import { JsonApiNetFilterExpressionVisitor, JsonApiNetPaginateExpressionVisitor } from '@dipscope/json-api-entity-provider';

import { ExchangeStore, MarketStore } from '.'

// Create entity provider.
const jsonApiExchangeProvider = new JsonApiEntityProvider({
  baseUrl: 'http://127.0.0.1:39006',
});

// Create entity store.
const exchangeStore = new ExchangeStore(jsonApiExchangeProvider);
const exchangeSet = exchangeStore.exchangeSet;

const run = async () => {
  // Find all exchanges
  await exchangeSet.findAll();
};

run();

```

**Comment by backnight** [comment]
I checked and my API endpoint `/exchanges` is requested, so the problem is happening after that

**Comment by dpimonov** [comment]
Hello,

Your code is correct! Can you also provide entity configurations and raw body response from a server? 

Want to verify the options and document format with a spec to be sure that there are no issues on a server side before digging further as tests in relation to JSON.NET backend runs successfully.

Thanks in advance!

**Comment by dpimonov** [comment]
Besides if you can provide project code for the server so I can run it locally it will be even more helpful.

**Comment by backnight** [comment]
ExchangeStore:
```
import { Type } from '@dipscope/type-manager';
import { EntitySet, EntityStore, EntityProvider } from '@dipscope/entity-store';

import { ExchangeModel } from './ExchangeModel';

@Type()
export class ExchangeStore extends EntityStore {
    public readonly exchangeSet: EntitySet<ExchangeModel>;

    public constructor(entityProvider: EntityProvider)
    {
        super(entityProvider);

        this.exchangeSet = this.createEntitySet(ExchangeModel);

        return;
    }
}
```


ExchangeModel:
```
import { Type, Property } from '@dipscope/type-manager';
import { JsonApiResource } from '@dipscope/json-api-entity-provider';

@Type()
@JsonApiResource({
  type: "exchanges"
})
export class ExchangeModel {
  @Property() public id: string;
  @Property() public name: string;

  constructor(id: string, name: string) {
    this.id = id;
    this.name = name;
  }
}
```


Raw response from server:

![Screenshot from 2022-08-15 16-25-25](https://user-images.githubusercontent.com/220051/184664899-1a83b493-b872-45e4-9ac7-f12621396e0e.png)
![Screenshot from 2022-08-15 16-25-36](https://user-images.githubusercontent.com/220051/184664981-eca452d5-9146-49e7-8a22-4181adf3ed2c.png)

**Comment by backnight** [comment]
Can't provide source easily, there is a ton of custom configuration for private js registry and more (docker, etc), I will do it if we can't find the issue

**Comment by dpimonov** [comment]
ExchangeStore:

```
import { EntitySet, EntityStore, EntityProvider } from '@dipscope/entity-store';
import { ExchangeModel } from './ExchangeModel';

export class ExchangeStore extends EntityStore 
{
    public readonly exchangeSet: EntitySet<ExchangeModel>;

    public constructor(entityProvider: EntityProvider)
    {
        super(entityProvider);

        this.exchangeSet = this.createEntitySet(ExchangeModel);

        return;
    }
}
```

`Type` decorator is not required here as you are not using this service for DI.

ExchangeModel:

```
import { Type, Property, Inject } from '@dipscope/type-manager';
import { JsonApiResource } from '@dipscope/json-api-entity-provider';

@Type()
@JsonApiResource({
  type: "exchanges"
})
export class ExchangeModel 
{
  @Property(Number) public id: number;
  @Property(String) public name: string;

  constructor(@Inject('id') id: string, @Inject('name') name: string) 
  {
    this.id = id;
    this.name = name;
  }
}
```

If you are not using `reflect-metadata` you have to explicitly define property types. To inject properties or services into constructor you have to use `Inject` decorator. If you have `reflect-metadata` included in your project or injection is not required you can declare your type as following:

```
import { Type, Property, Inject } from '@dipscope/type-manager';
import { JsonApiResource } from '@dipscope/json-api-entity-provider';

@Type()
@JsonApiResource({
  type: "exchanges"
})
export class ExchangeModel 
{
  @Property() public id: number;
  @Property() public name: string;
}
```

You can read [type manager docs](https://github.com/dipscope/TypeManager.TS#defining-decorators) for details. It is reflection and serialization core of entity store. But issue is not in the configuration. It is fully correct if you are using `reflect-metadata` library and string based id.

Your backend response is not correct in terms of [JSON:API](https://jsonapi.org/format/#fetching-resources) spec. That's why you got an error. Response should look like this in your case:

```
{
  "links": {
    "self": "http://127.0.0.1:39006/exchanges"
  },
  "data": [{
    "type": "exchanges",
    "id": "1",
    "attributes": {
      "name": "kraken"
    }
  }, {
    "type": "exchanges",
    "id": "2",
    "attributes": {
      "name": "ftx"
    }
  }]
}
```

I definitely have to provide a proper exception here if server response having wrong structure. Added to my TODO list. You have to check server configuration because now it seems to be incorrect if you are expecting JSON:API response.

**Comment by backnight** [comment]
Hello,

Thanks a lot for the feedbacks, I implemented everything and it's working now.

When calling:
```
const allExchanges = await exchangeSet.findAll();
console.log(allExchanges);
```

I get (and my data is in the "entities" field):
```
p {
  entities: [
    ExchangeModel { id: '1', name: 'kraken' },
    ExchangeModel { id: '2', name: 'ftx' }
  ],
  totalLength: undefined,
  typeMetadata: t {
    typeMetadataResolver: [Function: bound ],
    childrenTypeMetadataMap: Map(0) {},
    discriminantMap: Map(1) { [class ExchangeModel] => 'ExchangeModel' },
    propertyMetadataMap: Map(2) { 'id' => [t], 'name' => [t] },
    injectMetadataMap: Map(2) { 1 => [t], 0 => [t] },
    typeName: 'ExchangeModel',
    typeFn: [class ExchangeModel],
    typeFnMap: Map(0) {},
    typeOptionsBase: {
      discriminator: '$type',
      factory: e {},
      injector: [e],
      log: [e],
      preserveDiscriminator: false,
      referenceHandler: e {},
      serializer: e {},
      preserveNull: true,
      useDefaultValue: false,
      useImplicitConversion: false
    },
    typeOptions: {
      discriminant: 'ExchangeModel',
      propertyOptionsMap: [Map],
      injectOptionsMap: [Map],
      customData: [Object]
    },
    parentTypeMetadata: undefined
  },
  jsonApiAdapter: e {
    jsonApiConnection: e {
      baseUrl: 'http://127.0.0.1:39006',
      jsonApiRequestInterceptor: [Function (anonymous)],
      headers: [O [Headers]]
    },
    jsonApiMetadataExtractor: p {},
    allowToManyRelationshipReplacement: false
  },
  firstLinkObject: undefined,
  lastLinkObject: undefined,
  nextLinkObject: undefined,
  prevLinkObject: undefined
}
```

I understand the content, but I don't understand the "P" class which is containing everything. Is it an iterator?
Maybe you should give it another name than "P"

**Comment by backnight** [comment]
Okay it's a PaginatedEntityCollection, alright

**Comment by dpimonov** [comment]
Glad to hear that you've made it finally work.

> I understand the content, but I don't understand the "P" class which is containing everything. Is it an iterator?

Yeah as you have noticed this is `PaginatedEntityCollection`. It allows your to query next or prev chunks of data (if any) based on specified filter and sort conditions. There are plans to allow iteration in chunks to get the whole list when the data is paginated by a server and provide other benefits.

You can iterate a collection as a regular array. It will return data from current page or all entities if data is not paginated by a server.

```
const allExchanges = await exchangeSet.findAll();

for (const exchange of allExchanges)
{
    // Do stuff...
}
```

> Maybe you should give it another name than "P"

Yeah looks like I have to provide source maps as code is minified.

**Comment by dpimonov** [comment]
Bug is fixed in v1.0.1.
